### PR TITLE
Event-based block lookup tests

### PR DESCRIPTION
**Pull request by dapplion**
## Issue Addressed

Extends @jimmygchen refactor by
- Applying the changes to all existing block lookup tests
- Injects events directly on handle_message to avoid async machinery :)

This PR does not:
- Change the syntatic style of existing tests
- Modify or extend the tested scenarios. However it adds helper functions to make them more succint

Switching to event-based tests is very valuable to generalize tests, allow to test more diverse edge cases, and allow refactors without being too tied to block lookup internals.

This PR is an alternative to 
- https://github.com/sigp/lighthouse/pull/5530

## Proposed Changes

All tests now interface with sync through this event channels

```
                     +-----------------+
                     | BeaconProcessor |
                     +---------+-------+
                            ^  |
                            |  |
                  WorkEvent |  | SyncMsg
                            |  | (Result)
                            |  v
+--------+            +-----+-----------+             +----------------+
| Router +----------->|  SyncManager    +------------>| NetworkService |
+--------+  SyncMsg   +-----------------+ NetworkMsg  +----------------+
          (RPC resp)  |  - RangeSync    |  (RPC req)
                      +-----------------+
                      |  - BackFillSync |
                      +-----------------+
                      |  - BlockLookups |
                      +-----------------+
```

## Additional Info

RE: style we can discuss as follow-up to this PR, I see as the contentious options:

- Use declarative "state-less" API (pre-deneb tests)
- Use a chained builder-stype API for tests with internal implicit state (post-deneb)

Some notes to inform the above:

**chained API doesn't play well with backtrace**

Consider this code, and a panic in action_a

```
10: rig
11:   .action_a()
12:   .action_b()
13:   .action_a()
```

The backtrace line for `action_a` will declare line 10. You can't know which of the two action_a invocations caused the error

**internal implicit state succinctness**

Post-deneb test are shorter at the expense of clarity. There's machinery in the back that makes assumptions about what request is doing what. For straightforward cases it's fine but it can be restrictive for more complex cases. Pre-deneb tests with proper helper functions should look as succint as those with a chained API. We can remane `rig` to `r` and it almost looks like chained, which allowing conditional flow and storing variables.

**Comments**

Any sync tests are quite difficult to understand for those that have not written them. Having comments explaining the case is very important for either style. In that direction I think we should focus much more on having the test easy to understand and well documented with comments besides succintness

### CODE REVIEW

File: beacon_node/network/src/sync/block_lookups/mod.rs [hunk 1]
```diff
@@ -743,14 +764,19 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
...
         let request_state = R::request_state_mut(&mut lookup);
 
-        let Ok(peer_id) = request_state.get_state().processing_peer() else {
-            return;
+        let peer_id = match request_state.get_state().processing_peer() {
+            Ok(peer_id) => peer_id,
+            Err(e) => {
+                trace!(self.log, "Attempting to process single block lookup in bad state"; "id" => target_id, "response_type" => ?R::response_type(), "error" => e);
+                return;
+            }
```
Comment (dapplion) [inline beacon_node/network/src/sync/block_lookups/mod.rs]: Added this two traces to debug tests, as they were super helpful. This cases should never happen in practice and if they do, important events are ignored. I think they should be at least debug, thoughts?

Comment (realbigsean) [reply beacon_node/network/src/sync/block_lookups/mod.rs]: Yes, debug sounds good to me

File: beacon_node/network/src/sync/manager.rs [hunk 2]
```diff
@@ -619,6 +650,7 @@ impl<T: BeaconChainTypes> SyncManager<T> {
             SyncMessage::UnknownParentBlock(peer_id, block, block_root) => {
                 let block_slot = block.slot();
                 let parent_root = block.parent_root();
+                debug!(self.log, "received sync_message"; "message" => "UnknownParentBlock", "block_root" => %block_root, "parent_root" => %parent_root);
```
Comment (realbigsean) [inline beacon_node/network/src/sync/manager.rs]:
```suggestion
                debug!(self.log, "Received sync message"; "message" => "UnknownParentBlock", "block_root" => %block_root, "parent_root" => %parent_root);
```

File: beacon_node/network/src/sync/manager.rs [hunk 3]
```diff
@@ -638,6 +670,7 @@ impl<T: BeaconChainTypes> SyncManager<T> {
                 }
                 let mut blobs = FixedBlobSidecarList::default();
                 *blobs.index_mut(blob_index as usize) = Some(blob);
+                debug!(self.log, "received sync_message"; "message" => "UnknownParentBlob", "block_root" => %block_root, "parent_root" => %parent_root);
```
Comment (realbigsean) [inline beacon_node/network/src/sync/manager.rs]:
```suggestion
                debug!(self.log, "Received sync message"; "message" => "UnknownParentBlob", "block_root" => %block_root, "parent_root" => %parent_root);
```

File: beacon_node/network/src/sync/manager.rs [hunk 4]
```diff
@@ -646,11 +679,12 @@ impl<T: BeaconChainTypes> SyncManager<T> {
                     ChildComponents::new(block_root, None, Some(blobs)),
                 );
             }
-            SyncMessage::UnknownBlockHashFromAttestation(peer_id, block_hash) => {
+            SyncMessage::UnknownBlockHashFromAttestation(peer_id, block_root) => {
                 // If we are not synced, ignore this block.
                 if self.synced_and_connected(&peer_id) {
+                    debug!(self.log, "received sync_message"; "message" => "UnknownBlockHashFromAttestation", "block_root" => %block_root);
```
Comment (realbigsean) [inline beacon_node/network/src/sync/manager.rs]:
```suggestion
                    debug!(self.log, "Received sync message"; "message" => "UnknownBlockHashFromAttestation", "block_root" => %block_root);
```

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 5]
```diff
@@ -29,15 +31,44 @@ type T = Witness<ManualSlotClock, CachingEth1Backend<E>, E, MemoryStore<E>, Memo
...
+//                             |  v
+// +--------+            +-----+-----------+             +----------------+
+// | Router +----------->|  SyncManager    +------------>| NetworkService |
+// +--------+  SyncMsg   +-----------------+ NetworkMsg  +----------------+
+//           (RPC resp)  |  - RangeSync    |  (RPC req)
+//                       +-----------------+
+//                       |  - BackFillSync |
+//                       +-----------------+
+//                       |  - BlockLookups |
+//                       +-----------------+
```
Comment (realbigsean) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: Since half is doc format and half is comment format this looks weird in my IDE, not sure if this is intentional: 

Screenshot 2024-04-08 at 6 49 35 PM

```suggestion
///                      | BeaconProcessor |
///                      +---------+-------+
///                             ^  |
///                             |  |
///                   WorkEvent |  | SyncMsg
///                             |  | (Result)
///                             |  v
/// +--------+            +-----+-----------+             +----------------+
/// | Router +----------->|  SyncManager    +------------>| NetworkService |
/// +--------+  SyncMsg   +-----------------+ NetworkMsg  +----------------+
///           (RPC resp)  |  - RangeSync    |  (RPC req)
///                       +-----------------+
///                       |  - BackFillSync |
///                       +-----------------+
///                       |  - BlockLookups |
///                       +-----------------+
```

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 6]
```diff
@@ -56,91 +87,456 @@ impl TestRig {
...
+        let r = Self::test_setup();
+        if r.after_deneb() {
+            Some(r)
+        } else {
+            None
+        }
+    }
+
+    fn after_deneb(&self) -> bool {
+        matches!(self.current_fork(), ForkName::Deneb | ForkName::Electra)
```
Comment (realbigsean) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: Here's a place fork order would be useful, so we don't forget to enable these in future forks

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 7]
```diff
@@ -188,988 +601,557 @@ impl TestRig {
...
         if i < parent_lookup::PARENT_FAIL_TOLERANCE {
-            assert_eq!(
-                bl.parent_lookups[0]
-                    .current_parent_request
-                    .block_request_state
-                    .state
-                    .failed_attempts(),
-                dbg!(i)
-            );
+            // assert_eq!(
```
Comment (realbigsean) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: Are you able to add this back?

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 8]
```diff
@@ -188,988 +601,557 @@ impl TestRig {
...
```
Comment (realbigsean) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: This too

**Review by realbigsean — COMMENTED** [review]
Love the move to event based testing, thanks @jimmygchen and @dapplion !

The test do already look more straightforward and looks like there are some logging improvements in here. 

> stateless vs. builder-ish API

This PR is worth merging regardless of where we land there. 

> Any sync tests are quite difficult to understand for those that have not written them. Having comments explaining the case is very important for either style. In that direction I think we should focus much more on having the test easy to understand and well documented with comments besides succintness

Agree! My fault for lack of comments on a lot of the existing ones.

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 9]
```diff
@@ -56,91 +87,456 @@ impl TestRig {
...
+                .blob_ids
+                .to_vec()
+                .iter()
+                .all(|r| r.block_root == for_block) =>
+            {
+                Some(*id)
+            }
+            _ => None,
+        })
+        .expect(&format!("Expected blob parent request for {for_block:?}"))
```
Comment (realbigsean) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: There's a few instances of this failing the linter
```suggestion
        .unwrap_or_else(|_| panic!("Expected blob parent request for {for_block:?}"))
```

File: beacon_node/network/src/sync/block_lookups/tests.rs (hunk 7 above)
Comment (dapplion) [reply beacon_node/network/src/sync/block_lookups/tests.rs]: I don't think we should. The goal of this PR is to decouple tests from the internals of block lookups. This assertion is implicitly asserted in batch of the end of the test. What we really care about is that after N failures, the lookup is dropped, regardless of how accounting is done internally. I think I left the comment to double check with you, but I would vote to remove

File: beacon_node/network/src/sync/manager.rs [hunk 10]
```diff
@@ -233,31 +233,62 @@ pub fn spawn<T: BeaconChainTypes>(
...
+                beacon_processor.clone(),
+                beacon_chain.clone(),
+                log.clone(),
+            ),
+            range_sync: RangeSync::new(beacon_chain.clone(), log.clone()),
+            backfill_sync: BackFillSync::new(beacon_chain.clone(), network_globals, log.clone()),
+            block_lookups: BlockLookups::new(
+                beacon_chain.data_availability_checker.clone(),
+                log.clone(),
+            ),
```
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:272]: Unrelated to this PR but since you're touching this I thought it might be worth mentioning - perhaps we could create child logger for each of these individual components so we know where the log is coming from?

`log.new(slog::o!("service" => "range_sync")`

Comment (jimmygchen) [reply beacon_node/network/src/sync/manager.rs:272]: ps. i'm not suggesting a change in this PR

**Comment by dapplion** [comment]
Moved the logging changes into their own PR
- https://github.com/sigp/lighthouse/pull/5535

File: beacon_node/network/src/sync/block_lookups/tests.rs [hunk 11]
```diff
@@ -29,15 +31,44 @@ type T = Witness<ManualSlotClock, CachingEth1Backend<E>, E, MemoryStore<E>, Memo
 struct TestRig {
     beacon_processor_rx: mpsc::Receiver<WorkEvent<E>>,
     network_rx: mpsc::UnboundedReceiver<NetworkMessage<E>>,
+    network_rx_queue: Vec<NetworkMessage<E>>,
     rng: XorShiftRng,
     harness: BeaconChainHarness<T>,
+    sync_manager: SyncManager<T>,
+    beacon_processor: Arc<NetworkBeaconProcessor<T>>,
```
Comment (jimmygchen) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: Any reasons to have `beacon_processor` in the `TestRig`? I don't see it being used other than accessing `network_globals`.

I kept it out because the test rig should not need to directly interact with the `beacon_processor` directly in any way, otherwise it could interfere testing of the `SyncManager`.

Comment (jimmygchen) [inline beacon_node/network/src/sync/block_lookups/tests.rs]: Could we have some of the field documentation back as I think it would help people writing the tests understand what each of these fields are for, without reading through the implementation.

https://github.com/sigp/lighthouse/pull/5530/files#diff-13439a1c05bab56ff7fb8b91e2d4f569076456e8a909000e1d9d3b5df8afb1b0R59-R81

Comment (dapplion) [reply beacon_node/network/src/sync/block_lookups/tests.rs]: Added!

Comment (dapplion) [reply beacon_node/network/src/sync/block_lookups/tests.rs]: Good point! Removed beacon_processor for network_globals, and harness for fork_name

File: beacon_node/network/src/sync/manager.rs (hunk 10 above)
Comment (dapplion) [reply beacon_node/network/src/sync/manager.rs:272]: Does this label then apply to all logs? That's a good suggestion, can add to a future PR def

File: beacon_node/network/src/sync/block_lookups/tests.rs (hunk 7 above)
Comment (realbigsean) [reply beacon_node/network/src/sync/block_lookups/tests.rs]: let's remove 🚀

**Review by realbigsean — APPROVED** [review]

**Comment by realbigsean** [comment]
@mergifyio queue

**Comment by realbigsean** [comment]
@Mergifyio refresh

**Comment by realbigsean** [comment]
@mergifyio requeue

**Comment by jimmygchen** [comment]
@mergify requeue
